File: py/2025/5/1.py
def solve(input_data: str) -> str:
    p = input_data.strip().split("\n\n")
    r = sorted([tuple(map(int, l.split("-"))) for l in p[0].splitlines()])
    m = []
    for a, b in r:
        if not m or a > m[-1][1] + 1:
            m.append([a, b])
        else:
            m[-1][1] = max(m[-1][1], b)
    ids = list(map(int, p[1].split()))

    def f(x):
        left, right = 0, len(m) - 1
        while left <= right:
            mid = (left + right) // 2
            if m[mid][0] <= x <= m[mid][1]:
                return True
            if x < m[mid][0]:
                right = mid - 1
            else:
                left = mid + 1
        return False

    t = sum(f(x) for x in ids)
    return str(t)

File: rs/2025/src/5/1.rs
use std::io::{self, Read};

fn main() {
    let mut input = String::new();
    io::stdin()
        .read_to_string(&mut input)
        .expect("Failed to read input");

    let result = solve(&input.trim());
    println!("{}", result);
}

fn solve(input_data: &str) -> String {
    let parts: Vec<&str> = input_data.trim().split("\n\n").collect();
    let mut ranges: Vec<(i64, i64)> = parts[0]
        .lines()
        .map(|l| {
            let nums: Vec<i64> = l.split('-').map(|x| x.parse().unwrap()).collect();
            (nums[0], nums[1])
        })
        .collect();

    ranges.sort_by_key(|&(a, _)| a);

    let mut merged: Vec<[i64; 2]> = Vec::new();
    for (a, b) in ranges {
        if merged.is_empty() || a > merged.last().unwrap()[1] + 1 {
            merged.push([a, b]);
        } else {
            merged.last_mut().unwrap()[1] = merged.last().unwrap()[1].max(b);
        }
    }

    let ids: Vec<i64> = parts[1]
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();

    fn in_ranges(x: i64, merged: &[[i64; 2]]) -> bool {
        let mut left = 0;
        let mut right = merged.len() as isize - 1;
        while left <= right {
            let mid = (left + right) / 2;
            let mid_range = merged[mid as usize];
            if mid_range[0] <= x && x <= mid_range[1] {
                return true;
            } else if x < mid_range[0] {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        false
    }

    let t = ids.iter().filter(|&&x| in_ranges(x, &merged)).count();

    t.to_string()
}
